Sources/Build/ManifestBuilder.swift

```diff
@@ -362,92 +365,110 @@ extension LLBuildManifestBuilder {
-        collectTargetDependencyInfos(for: targetDescription.target,
-                                     targetDepGraphMap: targetDepGraphMap,
-                                     dependencyArtifactMap: &targetDependencyMap)
+        var dependencyModulePathMap: SwiftDriver.ExternalTargetModulePathMap = [:]
+        // Collect paths for target dependencies of this target (direct and transitive)
+        collectTargetDependencyModulePaths(for: targetDescription.target,
+                                           dependencyModulePathMap: &dependencyModulePathMap)
 
         // Compute the set of frontend
         // jobs needed to build this Swift target.
         var commandLine = targetDescription.emitCommandLine();
         commandLine.append("-driver-use-frontend-path")
         commandLine.append(buildParameters.toolchain.swiftCompiler.pathString)
         commandLine.append("-experimental-explicit-module-build")
-        // FIXME: At some point SwiftPM should provide its own executor for
-        // running jobs/launching processes during planning
         let resolver = try ArgsResolver(fileSystem: targetDescription.fs)
         let executor = SPMSwiftDriverExecutor(resolver: resolver,
                                               fileSystem: targetDescription.fs,
                                               env: ProcessEnv.vars)
         var driver = try Driver(args: commandLine, fileSystem: targetDescription.fs,
                                 executor: executor,
-                                externalModuleDependencies: targetDependencyMap)
+                                externalBuildArtifacts: (dependencyModulePathMap, discoveredModulesMap))
 
         let jobs = try driver.planBuild()
 
+        // Save the path to the target's module to be used by its dependents
         // Save the dependency graph of this target to be used by its dependents
         guard let dependencyGraph = driver.interModuleDependencyGraph else {
             fatalError("Expected module dependency graph for target: \(targetDescription)")
         }
-        targetDepGraphMap[targetDescription.target] = dependencyGraph
+        try accumulateDiscoveredModules(from: dependencyGraph,
+                                        discoveredModules: &discoveredModulesMap)
+
         try addSwiftDriverJobs(for: targetDescription, jobs: jobs, inputs: inputs, resolver: resolver,
                                isMainModule: { driver.isExplicitMainModuleJob(job: $0)})
     }
 
+    private func accumulateDiscoveredModules(
+        from dependencyGraph: InterModuleDependencyGraph,
+        discoveredModules: inout SwiftDriver.ModuleInfoMap
+    ) throws {
+        for (moduleId, moduleInfo) in dependencyGraph.modules {
+            switch moduleId {
+              case .swift:
+                discoveredModules[moduleId] = moduleInfo
+              case .clang:
+                if discoveredModules[moduleId] == nil {
```

Comment:
Hate to have my only comment be a coding-style one, but I think this would be cleaner as:

```swift
guard let existingModuleInfo = discoveredModules[moduleId] else {
  discoveredModules[moduleId] = moduleInfo
  break
}
InterModuleDependencyGraph.mergeClangModuleInfoDependencies(moduleInfo, existingModuleInfo)
discoveredModules[moduleId] = combinedDependenciesInfo
```

Reply:
That is a lot cleaner. Thank you!